### ISSUE TITLE
bundle to other data sources

### ISSUE THREAD

**Issue by berteh** [berteh/ScribusGenerator#10]
enable to use other data sources (DB, JSON) beside CSV.

**Comment by berteh** [comment]
not planned for now, but if it's important to you feel free to tell us, and help if you can ;)

**Comment by berteh** [comment]
+xml requested in #19

**Comment by berteh** [comment]
Good ways of exporting csv from postgresql and sqlite3 are now in readme:

**Mysql:**
```mysql --delimiter="," -u myuser -p mydb -e "select f1,f2 from mytable" > /tmp/mydata.txt```
or
```mysql  -u myuser -p  mydatabase -e 
    "select field1 , field2 FROM mytable INTO OUTFILE 
'/tmp/myfilename.csv' FIELDS TERMINATED BY ','
 ENCLOSED BY '\"' LINES TERMINATED BY '\n' " ```
moreover INTO OUTFILE at http://dev.mysql.com/doc/refman/5.1/en/select.html

**Postgresql:**
To create a CSV file from a Postgresql database:
toggle unaligned output with the ```\a``` switch, activate a comma as a separator with ```\f ,``` (without apostrophes). Send output to a file with ```\o myfile.csv```, then query your database.
 
**Sqlite3:**
You can use ```sqlite3 -csv``` in command line or ```.mode csv``` in sqlite's interactive shell

**Comment by S1SYPHOS** [comment]
Hey @berteh,
long time no see! As I had some time left after creating the 'reading tips' for our little children's bookstore, I had a look at your code and saw that you mentioned that implementing `csv.DictReader` could be nice. Then I thought, if we did this, parsing JSON isn't far around the corner, since the output CSV data would be a dictionary just like loading JSON would give me.

**TL;DR** I monkey-patched stuff together [here](https://github.com/S1SYPHOS/ScribusGenerator/tree/dict_json) which right now is far from perfect ofc since all code assumes that CSV was loaded row by row, requiring header row (= variable names) being passed and whatnot - **but I think this could be a good starting point**.

I created `examples/json1.sla` which is identical to `example/csv1.sla` and used `Business_Card.json` which is a JSON variant of `Business_Card.csv`.

I'd like to pursue this, but I feel like I need proper testing. What are the current requirements, Python 3.x? 2.7? How are you testing your code for several Scribus 1.4.x / 1.5.x versions?

Glad to hear from you, best regards
S1SYPHOS

**Comment by berteh** [comment]
Thanks @S1SYPHOS for the idea and effort.

I think this nice new feature could be targeted only at Scribus 1.5.6+, that runs Python3. No need to backport I think, since 1.5.6 is stable enough to use in production from my experience.

If you still decide to do a backport for the Scribus 1.4 branch (current stable), what I currently do is I have 2 branches of the same project... locally cloned into separate folders for dev and testing, as Github makes it easy enough to provide dynamic download links for branches:
https://github.com/berteh/ScribusGenerator/archive/python3.zip (scribus 1.5.6+, python 3)
https://github.com/berteh/ScribusGenerator/archive/master.zip  (older versions, python 2.7)

As soon as Scribus releases next stable (1.6?) I'd just update the branch names accordingly to make the python3 the new master.

For testing I run a few generations from command line and open them manually in appropriate Scribus for a quick chekc, but if you want to create a better non-regression testing means feel free... I just never go to that.

Still: feel free to do it any way you like, suggestions are welcome, and thanks again for the initiative !
B.

**Comment by berteh** [comment]
Quick look at your code looks fine to me. Thanks for cleaning some of my Dict enumerations too !

The python3 branch is at https://github.com/berteh/ScribusGenerator/tree/python3

**Comment by S1SYPHOS** [comment]
Thanks - I remember starting Python programming when I had to dive into Scribus for my first project, immediately turning to this very fine library. It's nice to give something back!

On topic: I just open a PR on the p3 branch and work on that, then?

**Comment by berteh** [comment]
Yes Martin, that would be perfect. Thank you.

**Comment by S1SYPHOS** [comment]
@berteh I'm on it - what version of python will Scribus 1.6 be running on? What should I use for testing?

**Comment by S1SYPHOS** [comment]
Just a(nother) quick question:
Why are conditions wrapped in braces, like `if(len(row) > 0):` .. instead of `if len(row) > 0:`? For more complex conditonals I get why this is useful, but in my example, I don't follow :fox_face:

**Comment by S1SYPHOS** [comment]
Update: PR #184 is open!
